### ISSUE TITLE
Progress reporter

### ISSUE THREAD

**Issue by richardwerkman** [stryker-mutator/stryker-net#9]
Right now the only status reporter is the dot reporter. It will print dots and characters after a mutant has been tested.

It would be nice to show users a progress bar to show the progress of the mutationtestrun. A estimation could also be displayed next to the progress bar.

**Comment by yaelkeemink** [comment]
Could be usefull to fix #105 first.

**Comment by richardwerkman** [comment]
@yaelkeemink why? I had imagined that the progress would be mutant based. Like 20/40 mutants tested = 50% done. We don't need the number of tests for that. Or did you have something else in mind?

**Comment by yaelkeemink** [comment]
Thats another way to aproach it. I thought it was based on the number of tests and the time needed to run the tests. that way we can try to predict how much time left till stryker is finished. Your idea it simpler to implement, so I suggest to implement your idea and extend it later.

**Comment by yaelkeemink** [comment]
@richardwerkman Do you agree with me that this is a good first issue then?

**Comment by richardwerkman** [comment]
Yes, this is a great first issue! To make my thoughts even more clear:

>that way we can try to predict how much time left till stryker is finished.

We still can predict the time left by using the past time. For example, stryker created 50 mutants and the initial testrun took 5s. An estimate could be stryker will run for ~50x5s. During the mutation testing we could refine this even more by measuring how much time the past mutants took to test. That way te predition will get better as the program runs.

So my suggestion is that the progress bar will show up after mutating and compiling. And only show during the testing of the mutants, since this part of stryker takes longest and is easiest to predict.

**Comment by Szpi** [comment]
I'm happy to take this issue :).

**Comment by Szpi** [comment]
I have implemented something like this. 
![1](https://user-images.githubusercontent.com/23508122/46576113-303e6000-c9c3-11e8-8e04-a695f1fadcaa.PNG)
![2](https://user-images.githubusercontent.com/23508122/46576115-33395080-c9c3-11e8-8d8c-6d40d602d175.PNG)
![3](https://user-images.githubusercontent.com/23508122/46576116-35031400-c9c3-11e8-9b55-96b2804e5175.PNG)
Should I add progress bar ? I though of something like [---------] -> [==-------] to visualize progress but that could be too much lines that are about the same thing. What do you think about it?

**Comment by Mobrockers** [comment]
The progress bar could be for example inside the percentage complete brackets like so [======52%==---------]

**Comment by richardwerkman** [comment]
I think it would be great if the progress reporter would be a substitute for the ConsoleDotReporter, and not an addition. Also the progress bar could use [special characters](https://en.wikipedia.org/wiki/Block_Elements#Character_table) to look better. Maybe something like this?
```
123 mutants have been created. Each mutant will now be tested, this could take a while.
| █████████-------- |  72/123  |  50% | ~2m15s |
Killed:    50
Survived:  12
Timed out: 11
```

**Comment by Mobrockers** [comment]
Did not feel like looking up the special characters on my phone last night but yea the black blocks is what I was going for 👍

**Comment by Szpi** [comment]
Current state of progress reporting: 
![new](https://user-images.githubusercontent.com/23508122/46629857-8f76ae80-cb42-11e8-839e-6c9f3304c895.PNG)
What do you think about it? I have removed ConsoleDotProgressReporter and replaced its functionality as described by @richardwerkman, therefore I think this solves also #146.

**Comment by richardwerkman** [comment]
This is looking awesome! Does it print this for every mutant? Or does it update the values? It would be really cool if it updates.

For the estimate, could the minutes count be left out when it reaches 0? And did you concider that times could also be hours? On really large projects.

**Comment by Szpi** [comment]
It updates fields. I will change time formatting, thank you for feedback :).

**Comment by richardwerkman** [comment]
> I have removed ConsoleDotProgressReporter and replaced its functionality

Oh, that was not what I mean't I'm sorry. I meant that we want to offer two reporters to users. The basic "DotReporter", and the more advanced "ProgressReporter". Users can configure this using the settings 
`--reporter progress` or `--reporter dot`. The dot reporter is nice for using on servers like azure, since there you won't need fancy status bars. The functionality for passing reporter parameters is already there. I'm sorry if I was not clear at first.

**Comment by Szpi** [comment]
Oh ok. Making settings will be cover in another issue?

**Comment by Szpi** [comment]
PR #175

**Comment by Mobrockers** [comment]
Closed by #175
